gpii/node_modules/eventLog/src/eventLog.js

```diff
@@ -0,0 +1,396 @@
+        level = null;
+    }
+    gpii.eventLog.startTimer(timerId);
+    gpii.eventLog.pendingLogs[timerId] = gpii.eventLog.createLogObject(module, event, data, level);
+};
+
+/**
+ * Logs a deferred entry, setting the duration field to the data object.
+ *
+ * @param timerId {String} The same timerId for the call to defer().
+ */
+gpii.eventLog.logDeferred = function (timerId) {
+    var eventObject = gpii.eventLog.pendingLogs[timerId];
+    if (eventObject) {
+        delete gpii.eventLog.pendingLogs[timerId];
+
+        gpii.eventLog.logDuration(eventObject, null, null, null, timerId);
+    } else {
+        fluid.log("logDeferred: Unknown timerId", timerId);
+
+    }
+};
+
+/**
+ * Logs an error.
+ *
+ * @param module {String} The part of GPII causing this error.
+ * @param errType {String} Type of error.
+ * @param err {Object} The error.
+ */
+gpii.eventLog.logError = function (module, errType, err) {
+    var data = {};
+    if (err instanceof Error) {
+        // Error doesn't serialise
+        data.error = {};
+        fluid.each(Object.getOwnPropertyNames(err), function (a) {
+            data.error[a] = err[a];
+        });
+    } else {
+        data.error = err;
+    }
+    gpii.eventLog.log(module, "Error." + errType, data, fluid.logLevel.FAIL);
+};
+
+// Log uncaught exceptions.
+fluid.onUncaughtException.addListener(function (err) {
+    gpii.eventLog.logError(null, "UncaughtException", err);
+}, "eventLog");
+
+// Log fluid.fail.
+fluid.failureEvent.addListener(function (args) {
+    var err = Array.isArray(args) ? args.join(" ") : err;
+    gpii.eventLog.logError(null, "Failure", err);
+}, "eventLog", "before:fail");
+
+/**
+ * Gets the path of the new log file for this instance of gpii.
+ *
+ * @return {String} Path of the log file.
+ */
+gpii.eventLog.getLogFilePath = function () {
+    if (!gpii.eventLog.logFilePath) {
+        var startupTime = Date.now();
+        gpii.eventLog.logFilePath = process.env.GPII_EVENT_LOG;
+
+        if (!gpii.eventLog.logFilePath) {
+            var settingsDirComponent = gpii.settingsDir();
+            var gpiiSettingsDir = settingsDirComponent.getGpiiSettingsDir();
+            gpii.eventLog.logFilePath =
+                path.join(gpiiSettingsDir, "gpii-" + gpii.journal.formatTimestamp(startupTime) + ".log");
+        }
+
+        fluid.log(fluid.logLevel.IMPORTANT, "Writing event log to " + gpii.eventLog.logFilePath);
+    }
+    return gpii.eventLog.logFilePath;
+};
+
+/**
+ * Writes an event to the log file.
+ *
+ * @param level {Object} Level of the log, see fluid.logLevelsSpec [FATAL,FAIL,WARN,IMPORTANT,INFO,TRACE].
+ * @param event {Object} The object.
+ */
+gpii.eventLog.writeLog = function (level, event) {
+    var intLevel = gpii.eventLog.checkLevel(level);
+    event.level = intLevel.value;
+
+    // Log to console before the installation ID and timestamp are added (no one wants to see it).
+    fluid.log(intLevel, event);
+
+    if (!gpii.eventLog.logFilePath) {
```

Comment:
Again stick all this mutable state into component members, and do the conditional logic on startup